8262520: Add SA Command Line Debugger support to connect to debug server

YaSuenag opened the pull request:
`attach` command on CLHSDB supports to attach live process (PID) and coredump, but it cannot connect to debug server. CLHSDB does not have a command to connect to debug server.

Other jhsdb commands (jstack, jmap, etc...) can connect debug server via `--connect` option, so CLHSDB should connect to it.

After this change, you can connect to debug server with 'connect <hostname>'.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8262520](https://bugs.openjdk.java.net/browse/JDK-8262520): Add SA Command Line Debugger support to connect to debug server


### Reviewers
 * [Chris Plummer](https://openjdk.java.net/census#cjplummer) (@plummercj - **Reviewer**)
 * [Kevin Walls](https://openjdk.java.net/census#kevinw) (@kevinjwalls - Committer)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2773/head:pull/2773`
`$ git checkout pull/2773`

plummercj commented:
You'll need a CSR for this.

Will `jhsdb clhsdb --connect debugserver` work? If not, it should to be consistent with the other commands.

YaSuenag commented:
> You'll need a CSR for this.

Ok, I will add it after this discussion ( `coonect` or `attach` )

> Will `jhsdb clhsdb --connect debugserver` work? If not, it should to be consistent with the other commands.

It does not work. CLHSDB and HSDB do not accept '--connect'. I plan to add it after this PR.

> Does `detach` undo the `connect`?

Yes.

> It's not very symetric if we do since we already had `attach` + `detach` and now you added `connect` + `detach`. Since you can `attach` to a pid or a core file, maybe instead of adding `connect` you should just put the debugd support in `attach`.

I thought `connect` command is easy to understand because it is similar with `--connect` option in jhsdb, and also the change is smaller than adding to `attach` because `attach` already has 1 string argument (PID).
I'm not familiar of network, but I concern hostname which is configured by numeric chars only - I'm not sure it is allowed, but [RFC 1123](https://tools.ietf.org/html/rfc1123) allows digit character at the first.

```
      The syntax of a legal Internet host name was specified in RFC-952
      [DNS:4].  One aspect of host name syntax is hereby changed: the
      restriction on the first character is relaxed to allow either a
      letter or a digit.  Host software MUST support this more liberal
      syntax.
```

If digit hostname (e.g. 1234) is allowed, we cannot add debugd support to `attach` because we cannot distinguish it is PID or hostname. If so, it is reasonable to add `connect` command for this purpose.

plummercj commented:
`jhsdb` has `--pid`, `--core`, and `--connect` as the 3 ways to "attach" for lack of a better word. It would have been nice if the clhsdb command did a better job of copying these three command line arguments. You are suggesting that adding a `connect` command would have similarity to the `--connect` option, but unfortunately the `attach` command has no such similarity with the command line arguments `--pid` and `--core`, so I'm not so sure doing so with `connect` is actually helping in that regard, especially when "connect" and "attach" pretty much mean the same thing.

As for numeric host names, yes, that is possible, but then you could also have a numeric core file name. If we really want to avoid the numeric host name problem, perhaps something like `attachd` would be better than `connect`, but it seems any choice we make will have it's drawbacks due the the baggage of existing commands and options.

YaSuenag commented:
@plummercj Ok, I try to add debug server support to `attach` clhsdb command. Then should we still need CSR?

YaSuenag commented:
> @plummercj Ok, I try to add debug server support to `attach` clhsdb command. Then should we still need CSR?

Pushed new commit to be implemented as `attach`. It works fine with serviceability/sa jtreg tests on my Linux x64.
I will add CSR if you are ok.

kevinjwalls commented:
Hi,
I ran the first version and agree it works, the connect command worked OK when I ran a local, manual test.  But I do like adding to the attach Command rather than connect, avoiding introducing a new Command name.

Assuming an int is a PID like in Tool.java is reasonable.  
You can have a leading numeral since a certain RFC, but although I can't find clear language to say it's definitely illegal, actualy using a fully numeric hostname seems unwise.

CLHSDB.java line 185
Now that "private void attachDebugger(int pid)" takes an int, we don't want to catch NumberFormatException, that try/catch block is not wanted.

In the testcase, I think it will fail if a command throws an Exception, but that is limiting what problems it can detect:
could we make the test check at least some of the key output to show that we are attached and getting information, not errors or something unexpected?  If at least some key commands we issue are checked that they contain some output indicating success that would be great.

YaSuenag commented:
Thanks @kevinjwalls !

> CLHSDB.java line 185
> Now that "private void attachDebugger(int pid)" takes an int, we don't want to catch NumberFormatException, that try/catch block is not wanted.

I fixed it in new commit, and also I fixed HSDB.java.

> In the testcase, I think it will fail if a command throws an Exception, but that is limiting what problems it can detect:
> could we make the test check at least some of the key output to show that we are attached and getting information, not errors or something unexpected? If at least some key commands we issue are checked that they contain some output indicating success that would be great.

I added calling `class java.lang.Object` and `class java.lang.String` in testcase. We can check whether the connection works fine through them.

plummercj commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/CLHSDB.java:
```diff
@@ -217,8 +225,9 @@ private void attachDebugger(final String executablePath, final String corePath)
         // Try to open this core file
         try {
-            System.err.println("Connecting to debug server, please wait...");
+            System.out.println("Connecting to debug server, please wait...");
             agent.attach(remoteMachineName);
+            this.remoteMachineName = remoteMachineName;
```
I'm a little surprised to see that there is already some support for connecting to a debug server in CLHSDB.java. Was it previously used from anywhere?

YaSuenag replied:
I cannot find out caller of `CLHSDB::connect`. I guess CLHSDB.java is copied from HSDB.java because they are similar - HSDB uses `connect()` to connect to debug server.

plummercj commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/HSDB.java:
```diff
@@ -422,17 +429,20 @@ public void run() {
+    // If execPath != null, it is the path of a jdk/bin/java
     // and coreFilename is the pathname of a core file we are
     // supposed to attach to.
+    // Finally, if remoteMachineName != null, we are supposed to connect to remote debug server.
```
Split into 2 lines.

plummercj commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/HSDB.java:
```diff
@@ -78,10 +78,10 @@ public static void main(String[] args) {
   private int pid;
   private String execPath;
   private String coreFilename;
+  private String remoteMachineName;
```
`debugServerName` or `remoteDebugServerName`

YaSuenag replied:
Rename to `debugServerName` in new commit.

plummercj commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/HSDB.java:
```diff
@@ -94,10 +94,19 @@ private void doUsage() {
+    pid = -1;
+    execPath = null;
+    coreFilename = null;
+    remoteMachineName = null;
```
In general it's not clear to me why HSDB.java needed to be modified in a way similar to CLHSDB.java. If the goal is really just to add debug server support to the `clshdb attach` command, why is HSDB involved in that? Or is this adding debug server support elsewhere also (in which case it's not clear to me where this exposed to the user)?

YaSuenag replied:
As I said in above, CLHSDB.java and HSDB.java are similar, and also I added `attach(String debugServerName)` to `CommandProcessor$DebuggerInterface` which is used in HSDB.java, so I changed HSDB.java even though it does not affect to HSDB.

We can implement `attach(String debugServerName)` as empty method, and it might be reasonable. What do you think?

plummercj replied:
I see now. Both HSDB and CLHSDB implement CommandProcessor$DebuggerInterface, and they pass an instance of this class to the CommandProcessor constructor, and that allows CommandProcessor to do callbacks for the attach (or connect). Given this, it looks like you do want to add support for all 3 attach modes to the HSDB CommandProcessor$DebuggerInterface implementation. Otherwise you won't be able to attach to a debug server when you bring up the command line support using the HSDB gui.

What's been confusing, and I'm finally starting to understand, is that CommandProcessor is the SA command line support, and clhsdb (CLHSDB) and hsdb (HSDB) both use CommandProcessor to provide the command line support to the user. Maybe you should re-purpose this CR/PR as "Add SA CommandProcessor support to connect to debug server", and make sure the CR talks about how this impacts both clhsdb and hsdb.

YaSuenag replied:
I've updated subject both CR and PR, and also I updated description in JBS.

plummercj reviewed (commented):
This pre-existing code in CLHSDB.java is a little concerning:

127     private void doUsage() {
128         System.out.println("Usage:  java CLHSDB [[pid] | [path-to-java-executable [path-to-corefile]] | help ]");
129         System.out.println("           pid:                     attach to the process whose id is 'pid'");
130         System.out.println("           path-to-java-executable: Debug a core file produced by this program");
131         System.out.println("           path-to-corefile:        Debug this corefile.  The default is 'core'");
132         System.out.println("        If no arguments are specified, you can select what to do from the GUI.\n");
133         HotSpotAgent.showUsage();
134     }

First concern is that you have not added "debug server" support to it.  2nd concern is the "If no arguments are specified, you can select what to do from the GUI" part. It looks like using the GUI you can launch the CLHSDB console, although I can't find in the source where/how that is actually done. In any case, i don't see any way to pass command arguments when you do that.

I'm just wondering how much of this CLHSDB support we want to keep beyond what the clhsdb command needs. Do we still want to support invoking CLHSDB.main() from the java command line?

YaSuenag commented:
> This pre-existing code in CLHSDB.java is a little concerning:
> 
> ```
> 127     private void doUsage() {
> 128         System.out.println("Usage:  java CLHSDB [[pid] | [path-to-java-executable [path-to-corefile]] | help ]");
> 129         System.out.println("           pid:                     attach to the process whose id is 'pid'");
> 130         System.out.println("           path-to-java-executable: Debug a core file produced by this program");
> 131         System.out.println("           path-to-corefile:        Debug this corefile.  The default is 'core'");
> 132         System.out.println("        If no arguments are specified, you can select what to do from the GUI.\n");
> 133         HotSpotAgent.showUsage();
> 134     }
> ```
> 
> First concern is that you have not added "debug server" support to it. 2nd concern is the "If no arguments are specified, you can select what to do from the GUI" part. It looks like using the GUI you can launch the CLHSDB console, although I can't find in the source where/how that is actually done. In any case, i don't see any way to pass command arguments when you do that.

I just added the feature to connect to debug server, so I did not change `CLHSDB(String[] args)` and help message.
I think it should be changed when we add `--connect` option to `jhsdb clhsdb`. (I will work for it after this PR)

> 
> I'm just wondering how much of this CLHSDB support we want to keep beyond what the clhsdb command needs. Do we still want to support invoking CLHSDB.main() from the java command line?

I think we should remove `main()` from CLHSDB/HSDB to be honest because they are called from SALauncher, and we cannot call them directly because they are not exported in module-info.

YaSuenag commented on src/jdk.hotspot.agent/doc/clhsdb.html:
```diff
@@ -36,7 +36,7 @@ <h3>Annotated output of CLHSDB help command</h3>
 Available commands:
   assert true | false <font color="red">turn on/off asserts in SA code</font>
-  attach pid | exec core  <font color="red">attach SA to a process or core</font>
+  attach pid | exec core | remote_server  <font color="red">attach SA to a process, core, or remote debug server</font>
```
Fixed it in new commit.

YaSuenag commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/CLHSDB.java:
```diff
@@ -42,10 +46,11 @@ public static void main(String[] args) {
+        // If execPath != null, it is the path of a jdk/bin/java
         // and coreFilename is the pathname of a core file we are
         // supposed to attach to.
+        // Finally, if remoteMachineName != null, we are supposed to connect to remote debug server.
```
Fixed it in new commit.

YaSuenag commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/CLHSDB.java:
```diff
@@ -107,10 +119,10 @@ public void reattach() {
     private int pid;
     private String execPath;
     private String coreFilename;
+    private String remoteMachineName;
```
Rename to `debugServerName` in new commit.

plummercj commented:
> I think we should remove `main()` from CLHSDB/HSDB to be honest because they are called from SALauncher, and we cannot call them directly because they are not exported in module-info.

Yes, I'm starting to come to the same conclusion. The only code that constructs a `CLHSDB` object, and therefore triggers the potential call to `doUsage()`, is `CLHSDB.main()`, and it is only called from `SALauncher` when using `jhsdb clhsdb`. We could get rid of `main()` and have some other entrypoint that is passed the values of all the possible options, rather than having to parse them in CLHSDB.java.

There is one test invokes java with the `CLHSDB` class. See SABase.java. It's part of ciReplay testing. I think it would be easily modified to just run `jhsdb clhsdb`

plummercj commented on src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/CLHSDB.java:
```diff
@@ -214,16 +223,17 @@ private void attachDebugger(final String executablePath, final String corePath)
         catch (DebuggerException e) {
             final String errMsg = formatMessage(e.getMessage(), 80);
-            System.err.println("Unable to connect to machine \"" + remoteMachineName + "\":\n\n" + errMsg);
+            System.err.println("Unable to connect to machine \"" + debugServerName + "\":\n\n" + errMsg);
```
"machine" -> "debug server"

YaSuenag replied:
I fixed it in new commit.

YaSuenag commented:
> Yes, I'm starting to come to the same conclusion. The only code that constructs a `CLHSDB` object, and therefore triggers the potential call to `doUsage()`, is `CLHSDB.main()`, and it is only called from `SALauncher` when using `jhsdb clhsdb`. We could get rid of `main()` and have some other entrypoint that is passed the values of all the possible options, rather than having to parse them in CLHSDB.java.
> 
> There is one test invokes java with the `CLHSDB` class. See SABase.java. It's part of ciReplay testing. I think it would be easily modified to just run `jhsdb clhsdb`

I think `debugd` and `jinfo` are also remove `main()` if they have because they do not provide commands like `jstack`.
I will file it to JBS, and will fix them later.

plummercj commented:
BTW, referring back to the change to have attach handle pids and debug servernames, and the concern about a host having a numeric name, we already have code that assumes we don't have to deal with this in Tools.java. I just stumbled across this:

```
           try {
              pid = Integer.parseInt(args[0]);
              debugeeType = DEBUGEE_PID;
           } catch (NumberFormatException e) {
              // try remote server
              remoteServer = args[0];
              debugeeType  = DEBUGEE_REMOTE;
           }
```

plummercj commented:
/csr needed

plummercj commented:
I think the changes look good now. Please create a CSR and then we can finish the review of this PR.

YaSuenag commented:
I filed CSR and finalized it: https://bugs.openjdk.java.net/browse/JDK-8263005

> BTW, referring back to the change to have attach handle pids and debug servernames, and the concern about a host having a numeric name, we already have code that assumes we don't have to deal with this in Tools.java. I just stumbled across this:

Thanks! I felt confident for current PR :)

plummercj commented on test/hotspot/jtreg/serviceability/sa/sadebugd/ClhsdbAttachToDebugServer.java:
```diff
@@ -0,0 +1,108 @@
+            out.stderrShouldBeEmptyIgnoreDeprecatedWarnings();
+            out.shouldMatch("^java/lang/Object @0x[0-9a-f]+$"); // for "class java.lang.Object"
+            out.shouldMatch("^java/lang/String @0x[0-9a-f]+$"); // for "class java.lang.String"
+            out.shouldHaveExitValue(0);
```
I think you should do more error checking here. If you look at ClhsdbLauncher you will see:
```
        // This will detect most SA failures, including during the attach.
        oa.shouldNotMatch("^sun.jvm.hotspot.debugger.DebuggerException:.*$");
        // This will detect unexpected exceptions, like NPEs and asserts, that are caught
        // by sun.jvm.hotspot.CommandProcessor.
        oa.shouldNotMatch("^Error: .*$");
```
I think you could use something similar here.

YaSuenag replied:
I added them in new commit.
BTW I found out to check warnings which relate to `-Xcheck:jni` at ClhsdbLauncher. I did not add them to ClhsdbAttachToDebugServer because they are not the issue which relates to debug server. Is it ok?

plummercj replied:
No need for the `-Xcheck:jni` checks. I added that recently because we did have a bug, but I think we get more than enough testing with it already when running the ClhsdbLauncher tests.
